---
title: "MulEA"
author: "Wiktor Jurkowski, Eszter Ari, David Fazekas, Cezary Turek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MulEA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global_options, include=TRUE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, error = TRUE)
```

This vignette was created during the development process.

MulEA (Multi Enrichment Analysis) is a package/tool for functional interpretation and integration of multi-omics data sets. It can work with ontologies such as GO, ... 

Implemented features include:

- [Mulea Database Workflow](#mulea-database-workflow)
    - [File](#file)
    - [Local Mulea DB](#local-mulea-db)
    - [Global Mulea DB](#global-mulea-db)
- [Data Analysis](#data-analysis)
    - [Analysis Methods](#analysis-methods)
        - [Contingency Table Base](#contingency-table-base)
            - [Hypergeometric Test](#hypergeometric-test)
            - [Fisher Test](#fisher-test)
            - [Chi Squared Test](#chi-squared-test)
        - [Ranking Base](#ranking-base)
            - [Kolmogorov Smirnov Test](#kolmogorov-smirnov-test)
        - [Graph Base](#graph-base)
    - [Extra Operations](#extra-operations)
        - [Multiple Comparisons Problem](#multiple-comparisons-problem)
            - [GSEA](#gsea)
            - [General Approach](#general-approach)
- [Code Snippet](#code-snippet)
- [Other](#other)


## Mulea Database Workflow
There are 3 ways to work with data in MulEA; directly from our global database, with a local database or local files. Detailed instructions follow.
see [the second section](#custom)  

### File
The file which includes the ontology must be in GMT: Gene Matrix Transposed file format (*.gmt) ([format explanation](http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29)). Examples of the file are under MulEA installation directory in the example folder. To find the installation directory use find.package("MulEA") command.

In order to read the GMT file as a dataframe use: `MulEA::readGmtFileAsDF()`. Method requires one (*) argument `gmtFilePath`.

- `gmtFilePath` - This is the path to the file, Example: `"R/MulEA/example/model.gmt"`

The ontology is then read into the data frame, such:
```{r, results = 'asis'} 
pathToModelGmtFile <- paste(find.package("MulEA"),"/example/model.gmt", sep = "")
modelDfFromFile <- MulEA::readGmtFileAsDF(gmtFilePath = pathToModelGmtFile)
knitr::kable(modelDfFromFile, caption = "Model Data Frame")
```

If you would like to save the data frame that represents the ontology of the GMT file, use: `MulEA::saveModelFromDataFrameToGmtFile()`. You have to provide two arguments `modelDF`, `gmtFilePath`.

- `modelDF` - ontology data frame which represents the GMT file.
- `gmtFilePath` - Path to file where you want to save the ontology. Example: `"R/MulEA/example/savedModel.gmt"`

```{r, eval=FALSE} 
   MulEA::saveModelFromDataFrameToGmtFile(modelDF = modelDfFromFile, gmtFilePath = pathToModelGmtFile)  

```
### Local Mulea DB
With MulEA you can have all your ontologys (files, global DB etc.) in one place. This place is the local DB. This allows an easy way to control, share and work with all your ontologys. Before storing or downloading ontologys, run the database using: `MulEA::startLocalDatabase()`, this method creates the same schema Global DB uses, but locally on your computer. If the schema has already been created it reads the database, one argument is required for this method:
- `muleaDBLocalization` - By default its value is `":memory:"` which means the DB is created in temporary memory. 


If you would like to save the database, for sharing purposes or later use, specificy a directory such: `MulEA::startLocalDatabase("/home/muleaDB/DB1")`. Which will save the database as `/home/muleaDB/DB1Mulea.sqlite`

To view the database install the Mozilla Firefox addon https://addons.mozilla.org/en-US/firefox/addon/sqlite-manager/ and then go Main Menu -> Database -> Connect Database, then choose the database file MulEA.sqlite.

It is good practise to close the database connection after you finish work with MulEA `MulEA::stopLocalDatabase()`.



If database is running you are ready to add the ontology to it: `MulEA::addModelToLocalDatabase()`.It accepts 7 arguments. Four of them are required (*) `model`, `taxonomy_id`, `model_source`, `version`, and other 3 are optional `scientific_name`, `common_english_name`, `description`.

```{r, eval=FALSE}
  MulEA::addModelToLocalDatabase(model = modelDfFromFile, taxonomy_id = 9001, model_source = "GO", version = 2);

```

- `model` - Data Frame with ontology For example a Data Frame returned by `MulEA::readGmtFileAsDF()`.
- `taxonomy_id` - Integer that represents a taxonomy id from NCBI. In case of homo sapiens use 9001.
- `model_source` -  String that represents the origin of the ontologoy. Gen Ontology can be "GO".
- `version` - Integer that represents the ontology number. It alows for multiple versions of the same ontology. If second ontology use 2.
- `scientific_name` - Text which represents the scientific name, aids in identifying the organism.
- `common_english_name` - Same as above
- `description` - Field for notes on the ontology

To list ontologys from you local DB invoke `MulEA::listModelsFromLocalDatabase()` (not implemented yet)

To remove an ontology from the database: `MulEA::removeModelFromLocalDatabase()`. Primary Key (PK) of organisms_models table as the argument.

- `taxonomy_id` - Integer represents the taxonomi id of organism.
- `model_source` - tring that represents the origin of the ontology. Gen Ontology can be "GO".
- `version` - Version of the ontology in DB.


To retrieve the ontology from the DB: `MulEA::getModelFromLocalDatabaseAsDf` or `MulEA::getModelFromLocalDatabaseAsList`. The data frame retrieval can be used with analysis methods provided by MulEA. The list representation can also be useful as input to other package's methods. Both versions require the same arguments, (*) `taxonomy_id`, `model_source`, `version`.

In addition MulEA provides a helper method to save a ontology from the local DB in GMT format. `MulEA::saveModelFromLocalDatabaseToFile()` with arguments `taxonomy_id`, `model_source`, `version`, `gmtFilePath` all of them are required (*).

- `gmtFilePath` - The path to gmt file where ontology from local DB will be saved.

### Global Mulea DB
(Not implemented yet.)


## Data Analysis
"Analysis of data is a process of inspecting, cleansing, transforming, and modeling data with the goal of discovering useful information, suggesting conclusions, and supporting decision-making.", [Wikipedia](https://en.wikipedia.org/wiki/Data_analysis). MulEA performs data analysis on data frames. We implemented three types of tests base on contingency tables, rankings and topology of data.


### Analysis Methods


#### Contingency Table Base
Hypothesis tests may be performed on contingency tables. Cells in contingency table are defined as relationships between the row and column variables. MulEA provides few tests which are based on a contingency table. Hypergeometric,  Fisher and Chi Squared tests are documented below.


##### Hypergeometric Test
Hypergeometric test bases on [hypergeometric distribution](https://en.wikipedia.org/wiki/Hypergeometric_distribution). To perform this: `MulEA::calculateHypergeometricTest()` with below arguments:

- `model` -  Read it from file or load from DBs.
- `sampleVector` - Vector of your experimental data. Example: `dataFromExperiment <- c("FBgn0004407", "FBgn0010438", "FBgn0003742")`.
- `adjustMethod = NA` - By default is is set to NA. You can specify an algorithm which helps you with [Multiple Comparisons Problem](#multiple-comparisons-problem).

Data frame which you achieve as a result of running these tests:

```{r, results = 'asis'} 
experimentalData <- c("FBgn0004407", "FBgn0010438", "FBgn0003742", "FBgn0029709", "FBgn0030341", "FBgn0037044", "FBgn0002887")
hypergeometricTest <- MulEA::calculateHypergeometricTest(model = modelDfFromFile, sampleVector = experimentalData)
knitr::kable(hypergeometricTest, caption = "Hypergeometric Test Result Data Frame")
```

Column names are `category`,	`description`,	`listOfValues`,	`listOfValuesUnderCategory`,	`q`,	`m`,	`n`,	`k`,	`p.value`. First three are copies from the input data frame. `listOfValuesUnderCategory` represents an intersection of the listvalues under category and experimental data vector. The contingency table in this test is represented by `q`,	`m`,	`n`,	`k`. P-values are in the last column of presented DF. 


##### Fisher Test
[Fisher test](#https://en.wikipedia.org/wiki/Fisher's_exact_test) is represented in MulEA by `MulEA::calculateFisherTest()`. It takes the same list of arguments as [Hypergeometric Test](#hypergeometric-test), but produces different results.

```{r, results = 'asis'} 
fisherTest <- MulEA::calculateFisherTest(model = modelDfFromFile, sampleVector = experimentalData)
knitr::kable(fisherTest, caption = "Fisher Test Result Data Frame")
```

The structure of the single cell test results column is the same as R documentation about  [Fisher's Exact Test](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/fisher.test.html). To retrieve from data from the frame use: `fisherTest[4,]$testResultsColumnName[[1]]`. It gets the fourth row of the data frame and takes an unlisted cell with test results such:

```{r, results = 'asis'} 
fisherTest[4,]$testResultsColumnName[[1]]
```


##### Chi Squared Test
The last test from that category is [Chi Squared Test](https://en.wikipedia.org/wiki/Chi-squared_test). To calculate it on the groups from the ontology data frame: `MulEA::calculateChiSquaredTest()`. It takes the same list of arguments as two previously presented methods. Results for the data frame shown below (First 3 rows).

```{r, results = 'asis'} 
chiSquaredTest <- MulEA::calculateChiSquaredTest(model = modelDfFromFile, sampleVector = experimentalData)
knitr::kable(head(chiSquaredTest, n = 3), caption = "Chi Squared Test Result Data Frame")
```

A Similar formula to this can be used to achive results for specific rows of data: `chiSquaredTest[2,]$testResultsColumnName[[1]]`. The result cell is in form of results from [Pearson's Chi-squared Test](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/chisq.test.html).

```{r, results = 'asis'} 
chiSquaredTest[4,]$testResultsColumnName[[1]]
```

#### Ranking Base
(not implemented yet)


##### Kolmogorov Smirnov Test
First implementation of ranking base tests is [Kolmogorov-Smirnov Test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test). To calculate that test we have to create a muleaKolmogorovSmirnovTest object. Signature of that object is "muleaKolmogorovSmirnovTest". This object accepts one constructor's argument testData. It is a list of ids. It represents ranking.

```{r, results = 'asis'} 
muleaPkgDir <- find.package("MulEA")
modelDfFromFile <- MulEA::readGmtFileAsDF(gmtFilePath = paste(muleaPkgDir,"/example/model.gmt", sep = ""))
```

```{r, results = 'asis', echo=TRUE} 
muleaDataObject <- new(Class = "muleaData", gmt = modelDfFromFile)
dataFromExperiment <- c("FBgn0004407", "FBgn0010438", "FBgn0003742", "FBgn0029709", "FBgn0030341", 
                        "FBgn0037044", "FBgn0002887", "FBgn0028434", "FBgn0030170", "FBgn0263831")
muleaKolmogorovSmirnovTestObject <- new("muleaKolmogorovSmirnovTest", testData = dataFromExperiment)
ksTestRes <- MulEA::runTest(muleaDataObject, muleaKolmogorovSmirnovTestObject)
```

Returned data frame from K-S test look like that:

```{r, results = 'asis'} 
knitr::kable(head(ksTestRes, n = 3), caption = "K-S Test Result Data Frame")
```
  
<br />
<br />
<br />
  
#### Graph Base {#graph-base}
(not implemented yet)

#### Test paragfaph {#custom}


### Extra Operations
(not implemented yet)


#### Multiple Comparisons Problem


##### GSEA
(not implemented yet)


##### General Approach
According to article [Multiple Comparisons Problem](https://en.wikipedia.org/wiki/Multiple_comparisons_problem) you can adjust p-values of tested ontologys by adding the `adjustMethod` argument to contingency table tests methods. The value of this argument is one of the listed values: "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr". References to this method can be found here: [link](http://stat.ethz.ch/R-manual/R-devel/library/stats/html/p.adjust.html).

Data frames with adjusted p-values contain one extra column:

```{r, results = 'asis'} 
hypergeometricTestWithAdjustment <- MulEA::calculateHypergeometricTest(model = modelDfFromFile, sampleVector = experimentalData, adjustMethod = "BH")
knitr::kable(hypergeometricTestWithAdjustment, caption = "Hypergeometric Test Result With Adjustment Data Frame")
```

## Code Snippet
You can copy and paste this code to start working with MulEA more easily. All Snippets of code regarding the vignette:

[Reading GMT File](#file).

`
pathToModelGmtFile <- paste(find.package("MulEA"),"/example/model.gmt", sep = "")
modelDfFromFile <- MulEA::readGmtFileAsDF(gmtFilePath = pathToModelGmtFile)
knitr::kable(modelDfFromFile, caption = "Model Data Frame")
`

[SQLite DB on local machine](#local-mulea-db).

`
creationOfLocalDB <- MulEA::startLocalDatabase("/home/koralgooll/doktorat/Rpackages/mulea/")
creationOfLocalDB <- MulEA::startLocalDatabase(":memory:")
`

`
stopDbResults <- MulEA::stopLocalDatabase()
`

`
MulEA::addModelToLocalDatabase(model = modelDfFromFile, taxonomy_id = 9001, model_source = "GO", version = 0)
modelDfFromLocalDB <- MulEA::getModelFromLocalDatabaseAsDf(taxonomy_id = 9001, model_source = "GO", version = 0)
modelListFromLocalDB <- MulEA::getModelFromLocalDatabaseAsList(taxonomy_id = 9001, model_source = "GO", version = 0)
MulEA::removeModelFromLocalDatabase(taxonomy_id = 9001, model_source = "GO", version = 0)
`

`
MulEA::saveModelFromDataFrameToGmtFile(modelDF = modelDfFromLocalDB, gmtFilePath = "/home/koralgooll/doktorat/Rpackages/mulea/Mulea/example/fromDf.gmt")
MulEA::saveModelFromLocalDatabaseToFile(taxonomy_id = 9001, model_source = "GO", version = 0, gmtFilePath = "/home/koralgooll/doktorat/Rpackages/mulea/Mulea/example/fromDb.gmt")
`

[Statistical tests base on contingency table](#contingency-table-base).

`
hTestResults <- MulEA::calculateHypergeometricTest(model = modelDfFromLocalDB, sampleVector = dataFromExperiment)
fTestResults <- MulEA::calculateFisherTest(model = modelDfFromFile, sampleVector = dataFromExperiment)
chTestResults <- MulEA::calculateChiSquaredTest(model = modelDfFromLocalDB, sampleVector = dataFromExperiment)
`

[Adjust P-values for Multiple Comparisons.](#multiple-comparisons-problem)

`
hTestResultsWithAdjustment <- MulEA::calculateHypergeometricTest(model = modelDfFromLocalDB, sampleVector = dataFromExperiment, adjustMethod = "BH")
fTestResultsWithAdjustment <- MulEA::calculateFisherTest(model = modelDfFromFile, sampleVector = dataFromExperiment, adjustMethod = "bonferroni")
chTestResultsWithAdjustment <- MulEA::calculateChiSquaredTest(model = modelDfFromLocalDB, sampleVector = dataFromExperiment, adjustMethod = "BY")
`


## Other
(not implemented yet)

