---
title: "MulEA"
author: "Wiktor Jurkowski, Eszter Ari, David Fazekas, Cezary Turek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MulEA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global_options, include=TRUE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, error = TRUE)
```

THIS IS VIGNETTE in time of development. :)

Mulea is a package to acheave models of ... from our global DB or give you possibility
to put your own model to your local database. Of course you can push it to global DB.
Then you can use implemented mthods to alalysis data.

Actually implemented features (explained in details ... ) are:

- [Mulea Database Workflow](#mulea-database-workflow)
    - [File](#file)
    - [Local Mulea DB](#local-mulea-db)
    - [Global Mulea DB](#global-mulea-db)
- [Data Analysis](#data-analysis)
    - [Analysis Methods](#analysis-methods)
        - [GSEA](#gsea)
        - [Hypergeometric Test](#hypergeometric-test)
        - [Fisher Test](#fisher-test)
        - [Chi Squared Test](#chi-squared-test)
        - [Kolmogorov Smirnov Test](#kolmogorov-smirnov-test)
    - [Extra Operations](#extra-operations)
        - [Multiple Comparisons Problem](#multiple-comparisons-problem)
- [Other](#other)



## Mulea Database Workflow
We have 3 ways of working with data in MulEA package. You can work directly from our global DB, you can work with your local DB or with files. Below you can find detailed instructions.

### File
File which includes model need to be in GMT: Gene Matrix Transposed file format (*.gmt) ([format explanation](http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29)). You can find example of that file under MulEA installation directory in example folder. To find installation directory use find.package("MulEA") command.

To read GMT file as dataframe use this methos please: `MulEA::readGmtFileAsDF()`. Method accepts one required (*) argument `gmtFilePath`.

- `gmtFilePath` - It is path to file, which include model. Example: `"R/MulEA/example/model.gmt"`

Model is read into the data frame which looks like that:

```{r, results = 'asis'} 
pathToModelGmtFile <- paste(find.package("MulEA"),"/example/model.gmt", sep = "")
modelDfFromFile <- MulEA::readGmtFileAsDF(gmtFilePath = pathToModelGmtFile)
knitr::kable(modelDfFromFile, caption = "Model Data Frame")
```

### Local Mulea DB
With MulEA you can have all your models (models from files, global DB etc.) in one place. This place is local DB. This open for you an easy way to control, share and work with all your models. Before storing or downloading models thare you need to run database. To do it please use `MulEA::startLocalDatabase()`, this method creates the same schema as on Global DB, but locally on your computer. If schema has been already created it reads database. This method also has one argument:

- `muleaDBLocalization` - By default its value is `":memory:"` which means that DB is created in memory of computer. If you want to have file which represents you database, for example for sharing it, you should put a directory when SQLite file should be created. Example: `"/home/muleaDB/"`. File with name `MulEA.sqlite` will be created there.

To see how database look like please install add-on to Mozilla Firefox https://addons.mozilla.org/en-US/firefox/addon/sqlite-manager/ and then go to url chrome://sqlitemanager/content/sqlitemanager.xul choose from main menu Database > Connect Database, then choose file MulEA.sqlite.

It is good practise to close database connection after you finish work with MulEA. To do it type that command: `MulEA::stopLocalDatabase()`.

Then you can add models from your files using `MulEA::addModelToLocalDatabase()` function. In my case usage of function look like that: `MulEA::addModelToLocalDatabase("/home/koralgooll/doktorat/Rpackages/mulea/Mulea/example/REACTOME_database_dmel_uniq_PROPER.txt", taxonomy_id = 9006, model_source = 'File', version = 0, scientific_name = "realanimal", description = 'Test on real data')`. 

What file formats are supported? It is only GMT file format. All informatin about that format you can find at http://www.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29 . File has to include at teast three columns (collection_id, collection_name, collection) separated by TAB. Of course you can have more columns, because collection can have from 1 to X columns.

Example of model file:  

REACT_100100	Organic_anion_transporters	FBgn0027610  
REACT_100129	Ribosomal_scanning_and_start_codon_recognition	FBgn0001942	FBgn0002590	FBgn0002622	FBgn0003517	FBgn0003942	FBgn0004403	FBgn0004404	FBgn0004867	FBgn0004926	FBgn0005533	FBgn0010198	FBgn0010265	FBgn0010408	FBgn0010411	FBgn0010412	FBgn0011284	FBgn0015218	FBgn0015521	FBgn0015834	FBgn0017545  
REACT_100487	Mitochondrial_tRNA_aminoacylation	FBgn0016687	FBgn0020766	FBgn0021750	FBgn0027081	FBgn0027082	FBgn0027083	FBgn0027084	FBgn0027085	FBgn0027087	FBgn0027088	FBgn0027090	FBgn0033900  
REACT_100693	FGFR3_ligand_binding_and_activation	FBgn0005592	FBgn0010389	FBgn0014135  

taxonomy_id, model_source, version are primary keays, so you have to type this arguments, others are not obligatory. Each addition will create model table with row in organisms_models table. Model table name is concatenation of "model" + taxonomy_id + model_source + version. 

You can remove model from DB by `MulEA::removeModelFromLocalDatabase(taxonomy_id = 9016, model_source = 'File', version = 0)` or update it by putting upper version integer.

You can save model from your local DB to file by `MulEA::saveModelFromLocalDatabaseToFile(filePath = "/home/koralgooll/doktorat/Rpackages/mulea/Mulea/example/savedModel.txt", taxonomy_id = 9006, model_source = 'File', version = 0)`

### Global Mulea DB
(Not implemented yet.)


## Data Analysis

### Analysis Methods

#### GSEA

GSEA is Gene Set Enrichment Analisys. 

Using hypergeometric formula to calculate enrichment of certain genes in a differential expression setup.

Hypergeometric GSEA is a simple command line program written in C++ which calculates the hypergeometric probability of different gene sets.

To run that implemantation on data just use `MulEA::runGSEA)` method. In my case it looks like that: `MulEA::runGSEA(databaseFilePath = "/home/koralgooll/experiments/GSEA/C/GSEA/Database/small_db.txt", populationFilePath = "/home/koralgooll/experiments/GSEA/C/GSEA/Database/small_pool.txt", sampleFilePath = "/home/koralgooll/experiments/GSEA/C/GSEA/Database/small_sample.txt")`
For now it only works with file, in future it will work directly on DB. So to use it you have to export model from local DB by MulEA::saveModelFromLocalDatabaseToFile() to small_db.txt. Other two files have formats:

small_pool.txt  
1  
2  
3  
4  
5  

small_sample.txt  
1  
2  
3  

where exported file looks like that:
small_db.txt  
category1 1 2	3	4  
category2	5	6	7	8  

#### Hypergeometric Test

Not implemented yet.

#### Fisher Test

Not implemented yet.

#### Chi Squared Test

TODO

#### Kolmogorov Smirnov Test

### Extra Operations

#### Multiple Comparisons Problem

## Other
